README.md
# chkconfig | [![CI Status](https://github.com/fedora-sysv/chkconfig/workflows/Integration%20tests/badge.svg)](https://github.com/fedora-sysv/chkconfig/actions?query=workflow%3AIntegration+tests) [![CodeQL](https://github.com/fedora-sysv/chkconfig/actions/workflows/codeql.yml/badge.svg)](https://github.com/fedora-sysv/chkconfig/actions/workflows/codeql.yml)


changelog
* Wed Oct 05 2022 Jan Macku <jamacku@redhat.com> - 1.21-1
- ci: Add CodeQL to replace LGTM
- alternatives: replace master/slave with leader/follower
- chkconfig: use correct cmp function
- Bump redhat-plumbers-in-action/differential-shellcheck from 2 to 3
- ci: Add Shell linter - Differential ShellCheck
- ci: Use more inclusive terminology in workflows
- ci: Update workflows, packit and dependabot
- Translated using Weblate (Friulian)
- Translated using Weblate (Swedish)
- Translated using Weblate (Estonian)
- Translated using Weblate (Georgian)
- Translated using Weblate (Polish)
- Translated using Weblate (Korean)
- Translated using Weblate (Czech)
- Translations update from Fedora Weblate (#77)
- Translations update from Fedora Weblate (#75)
- Translations update from Fedora Weblate (#74)
- Translations update from Fedora Weblate (#73)
- Translated using Weblate (Ukrainian)
- Update translation files
- Family mentioned for --set in both man and help
- Translated using Weblate (French)
- build-sys: Ensure `systemd-sysv-install` symlink does not have `//`
- Translated using Weblate (German)
- Add LGTM badges to README
- Merge remote-tracking branch 'weblate/master'
- Translated using Weblate (Indonesian)
- Translated using Weblate (Finnish)
- Translated using Weblate (Korean)
- Translated using Weblate (Ukrainian)
- Translated using Weblate (Turkish)
- Translated using Weblate (Polish)
- Translated using Weblate (Norwegian Nynorsk)
- Update translation files
- Translated using Weblate (Finnish)
- Translated using Weblate (Czech)
- Translated using Weblate (Swedish)
- Translated using Weblate (Italian)
- Translated using Weblate (Spanish)
- Translated using Weblate (Chinese (Simplified))

* Wed Jul 28 2021 Jan Macku <jamacku@redhat.com> - 1.20-1
- spec: Replace not working awk command with sed (#62)

* Fri Jul 23 2021 Jan Macku <jamacku@redhat.com> - 1.19-1
- spec: Add Provides /sbin/chkconfig in order to stay backwards compatible (#60)

* Fri Jul 23 2021 Jan Macku <jamacku@redhat.com> - 1.18-1
- spec: /sbin/chkconfig -> /usr/sbin/chkconfig (#59)

* Thu Jul 22 2021 Jan Macku <jamacku@redhat.com> - 1.17-1
- alternatives: tweak manpage to match the real 'remove' behavior (#58)

* Thu Jul 15 2021 Jan Macku <jamacku@redhat.com> - 1.16-1
- alternatives: add --keep-foreign (#57)
- Translations update from Weblate
- ci: Onboard chkconfig to Packit
- zanata: remove zanata related stuff
- Use make macros
- alternatives: use one function for path cleaning
- CI: specify more closely when to run CI
- Add basic CI and README
- spec: sync specfile with Fedora

* Thu Jan 21 2021 Jan Macku <jamacku@redhat.com> - 1.15-1
- spec: sync specfile with Fedora
- makefile: Use rpmdev-bumpspec's legacy date option
- Add feature to generate specfile entry, commit and archive

* Tue Jul 21 2020 Tom Stellard <tstellar@redhat.com> - 1.14-2
- Use make macros
- https://fedoraproject.org/wiki/Changes/UseMakeBuildInstallMacro

* Fri Jul 17 2020 Jan Macku <jamacku@redhat.com> - 1.14-1
- Fix spelling of SELinux
- Remove hardcoded systemd path

* Tue Apr 14 2020 Jan Macku <jamacku@redhat.com> - 1.13-1
- fix typo in translations and fix bogus dates in changelog

* Mon Apr 06 2020 Jan Macku <jamacku@redhat.com> - 1.12-1
- alternatives setService(): Add missing error mesg - (#1820089)
- po: update translations
- rebase

* Thu Mar 14 2019 Peter Robinson <pbrobinson@fedoraproject.org> 1.11-4
- Split out alternatives into it's own package

* Mon Oct 08 2018 Lukas Nykryn <lnykryn@redhat.com> - 1.11-2
- add Provides: alternatives

* Mon Sep 10 2018 Lukas Nykryn <lnykryn@redhat.com> - 1.11-1
- Add tests for --add/remove-slave and use beakerlib
- alternatives: add-slave and remove-slave
- leveldb: don't crash on long names
- alternatives: prettier --list output

* Fri Apr 21 2017 Lukáš Nykrýn <lnykryn@redhat.com> - 1.10-1
- Introduce --remove-all option
- po: update translations

* Fri Feb 24 2017 Lukáš Nykrýn <lnykryn@redhat.com> - 1.9-1
- move sources to github

* Wed Jun 29 2016 Lukáš Nykrýn <lnykryn@redhat.com> - 1.8-1
- alternatives: introduce --keep-missing
- alternatives: allow family in --set and display it in --config
- chkconfig: use isXinetdEnabled instead of isOn
- leveldb: trim leading whitespaces from systemctl
- leveldb: suppress error messages when selinux is turned off
- alternatives: always recreate symlinks when the alternative is updated
- test-alternatives: basic tests for slave links
- chkconfig: resetpriorities should work on all runlevels

* Tue Nov 24 2015 Lukáš Nykrýn <lnykryn@redhat.com> - 1.7-1
- leveldb: fix segfault when selinux policy is not present
- alternatives: add family option

* Fri Oct 02 2015 Lukáš Nykrýn <lnykryn@redhat.com> - 1.6-1
- systemd-sysv-install: don't play ping-pong with systemctl
- ntsysv: add description to systemd services
- ntsysv: skip templates
- Makefile: fix typo

* Mon Jun 01 2015 Lukáš Nykrýn <lnykryn@redhat.com> - 1.5-1
- add systemd-sysv-install alias
- don't create symlinks if they already exist
- fix wrongly behaving LDFLAGS

* Thu Mar 26 2015 Lukáš Nykrýn <lnykryn@redhat.com> - 1.4-1
- ntsysv: show systemd services and sockets
- fix combination --type xinetd --list service
- leveldb: restore selinux context for xinetd conf files
- alternatives: remove unused variable
- alternatives: warn if the target is not a symlink
- spec: add link to git
- lets simplify version

* Wed Nov 05 2014 Lukáš Nykrýn <lnykryn@redhat.com> - 1.3.63-1
- alternatives: during install don't call preset on enabled services

* Tue Aug 12 2014 Lukáš Nykrýn <lnykryn@redhat.com> - 1.3.62-1
- use systemctl preset, not systemctl enable
- fix typo in manpage
- partly support socket activated services

* Wed Jul 31 2013 Lukáš Nykrýn <lnykryn@redhat.com> - 1.3.61-1
- try to make install_initd work
- fix permission issues with xinetd services

* Tue Mar 12 2013 Lukáš Nykrýn <lnykryn@redhat.com> - 1.3.60-1
- don't completely override LDFLAGS
- pass along any rpm-configured LD flags

leveldb.h
/* Copyright 1997-2008 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */
#ifndef H_LEVELDB
#define H_LEVELDB

#define RUNLEVELS "/etc"
#define XINETDDIR "/etc/xinetd.d"

#include <glob.h>

#define TYPE_INIT_D 0x1
#define TYPE_XINETD 0x2
#define TYPE_SYSTEMD 0x4
#define TYPE_ANY (TYPE_INIT_D | TYPE_XINETD | TYPE_SYSTEMD)

#ifndef SYSTEMD_SERVICE_PATH
#define SYSTEMD_SERVICE_PATH "/lib/systemd/system"
#endif

#ifndef SYSTEMD_LOCAL_SERVICE_PATH
#define SYSTEMD_LOCAL_SERVICE_PATH "/etc/systemd/system"
#endif

struct dep {
    char *name;
    int handled;
};

struct service {
    char *name;
    int levels, kPriority, sPriority;
    int currentLevels;
    char *desc;
    struct dep *startDeps;
    struct dep *stopDeps;
    struct dep *softStartDeps;
    struct dep *softStopDeps;
    char **provides;
    int type;
    int isLSB;
    int enabled;
};

int parseLevels(char *str, int emptyOk);

/* returns 0 on success, 1 if the service is not chkconfig-able, -1 if an
   I/O error occurs (in which case errno can be checked) */
int readServiceInfo(char *name, int type, struct service *service,
                    int honorHide);
int readServices(struct service **services);
int readServiceDifferences(char *name, int type, struct service *service,
                           struct service *service_overrides, int honorHide);
int parseServiceInfo(int fd, char *name, struct service *service, int honorHide,
                     int partialOk);
int currentRunlevel(void);
int isOn(char *name, int where);
int isConfigured(char *name, int level, int *priority, char *type);
int whatLevels(char *name);
int doSetService(struct service s, int level, int on);
int findServiceEntries(char *name, int level, glob_t *globresptr);
int readXinetdServiceInfo(char *name, struct service *service);
int setXinetdService(struct service s, int on);
int systemdIsInit();
int systemdActive();
int isOverriddenBySystemd(const char *service);
int isEnabledInSystemd(const char *service);
int isSocketActivatedBySystemd(const char *service);
int readSystemdUnitProperty(char *name, char *property, char **value);

void checkSystemdDependencies(struct service *s);
#endif


test-alternatives.sh
#!/bin/bash

# Include the BeakerLib environment
. /usr/share/beakerlib/beakerlib.sh

# Set the full test name
TEST="Test Alternatives"

# Package being tested
PACKAGE="chkconfig"

# We need to test both new "leader/follower" and legacy "master/slave" options
FOLLOWER_OR_SLAVE="follower"

function clean_dir {
    [ -n ${altdir} -a -d ${altdir} ] && rm ${altdir}/* &> /dev/null
    [ -n ${admindir} -a -d ${admindir} ] && rm ${admindir}/* &> /dev/null

    if [ -n ${testdir} -a -d ${testdir} ] ; then
        for i in ${testdir}/* ; do
            [ -f ${i} -o -L ${i} ] &&  rm ${i}
            [ -d ${i} ] && ( rm ${i}/* ; rmdir ${i} )
        done
    fi
}

function add_alternative {
    path="${testdir}/$1/main"
    prio=$2
    family=
    follower=follower
    mkdir -p "${testdir}/$1"
    touch $path

    [ -n "$3" ] && family="--family $3"

    [ -n "$4" ] && follower=${4}

    spath="${testdir}/$1/$follower"
    touch ${spath}

    rlRun "./alternatives --altdir ${altdir} --admindir ${admindir} --install ${link} ${name} ${path} ${prio} --${FOLLOWER_OR_SLAVE} ${slink} ${sname} ${spath} ${family}" 0 "NEW\tlink: $1\tPrio: $prio\tFamily: $3"
}

function remove_alternative {
    path="${testdir}/$1/main"
    rm ${testdir}/$1/*
    rmdir ${testdir}/$1

    rlRun "./alternatives --altdir ${altdir} --admindir ${admindir} --remove ${name} ${path}" 0 "REMOVE\tlink: $1"
}

function set_alternative {
    path="${testdir}/$1/main"

    rlRun "./alternatives --altdir ${altdir} --admindir ${admindir} --set ${name} ${path}" 0 "SET\tlink: $1"
}

function add_follower {
    path="${testdir}/$1/main"
    follower=follower
    touch $path

    [ -n "$2" ] && follower=${2}

    spath="${testdir}/$1/$follower"
    touch ${spath}
    rlRun "./alternatives --altdir ${altdir} --admindir ${admindir} --add-${FOLLOWER_OR_SLAVE} ${name} ${path} ${slink} ${sname} ${spath}" 0 "NEW_FOLLOWER\tlink: $spath"
}

function remove_follower {
    path="${testdir}/$1/main"
    follower=follower
    touch $path

    [ -n "$2" ] && follower=${2}
    rlRun "./alternatives --altdir ${altdir} --admindir ${admindir} --remove-${FOLLOWER_OR_SLAVE} ${name} ${path} ${sname}" 0 "NEW_FOLLOWER\tlink: $spath"
}

function check_alternative {
    path=$1
    shift
    state=$1
    shift
    follower=follower

    if [ "$state" = "manual" ] ; then
         best=${1}
         shift
    else
         best=${path}
    fi

    if [ "$1" = EMPTY ] ; then
        follower=
    elif [ -n "$1" ] ; then
        follower=$1
    fi

    cur_path=$(readlink ${altdir}/${name} | xargs dirname | xargs basename)
    cur_state=$(head -1 ${admindir}/${name})
    cur_best=$(LC_ALL=C ./alternatives --altdir ${altdir} --admindir ${admindir} --display TEST | grep best | cut -d " " -f5 | sed -e 's/\.$//' | xargs dirname | xargs basename)
    cur_spath=$(readlink ${altdir}/${sname} | xargs basename)
    echo $cur_spath
    rlAssertEquals "Mode:" "${state}" "${cur_state}"
    rlAssertEquals "Highest Priority:" "${best}" "${cur_best}"
    rlAssertEquals "Selected:" "${path}" "${cur_path}"
    rlAssertEquals "Follower:" "${cur_spath}" "${follower}"
}

rlJournalStart
    # Setup phase: Prepare test directory
    rlPhaseStartSetup
        rlAssertRpm $PACKAGE
        rlRun 'altdir=$(mktemp -d)' 0 'Creating tmp directory' # no-reboot
        rlRun 'admindir=$(mktemp -d)' 0 'Creating tmp directory' # no-reboot
        rlRun 'testdir=$(mktemp -d)' 0 'Creating tmp directory' # no-reboot

        rlRun 'name="TEST"'
        rlRun 'link="${testdir}/main_link"'

        rlRun 'sname="STEST"'
        rlRun 'slink="${testdir}/follower_link"'
    rlPhaseEnd

    for n in "follower" "slave"  ; do

    FOLLOWER_OR_SLAVE=$n
        # Test phase: Testing touch, ls and rm commands
        rlPhaseStart FAIL "Create Alternative"
            add_alternative link_a 10 ""
            check_alternative link_a auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Set Manual"
            add_alternative link_a 10 ""
            set_alternative link_a
            check_alternative link_a manual link_a
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Auto Priority Ascendant"
            add_alternative link_a 10 ""
            add_alternative link_b  20
            check_alternative link_b auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Auto Priority Descendant"
            add_alternative link_a 20 ""
            add_alternative link_b 10 ""
            check_alternative link_a auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Manual Overrides Best"
            add_alternative link_a 10 ""
            set_alternative link_a
            add_alternative link_b  20 ""
            check_alternative link_a manual link_b
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Remove Manually Set"
            add_alternative link_a 10 ""
            set_alternative link_a
            add_alternative link_b  20 ""
            remove_alternative link_a
            check_alternative link_b auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Follower"
            add_alternative link_a 10 "" follower_a
            add_alternative link_a 10 "" follower_b
            check_alternative link_a auto follower_b
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Follower Manual"
            add_alternative link_a 10 "" follower_a
            set_alternative link_a
            add_alternative link_a 10 "" follower_b
            check_alternative link_a manual link_a follower_b
            clean_dir
        rlPhaseEnd
            
            ##########
            
        rlPhaseStart FAIL "Family Priority Ascendant"
            add_alternative link_a 10 family_a
            add_alternative link_b 20 family_a
            check_alternative link_b auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Family Priority Descendant"
            add_alternative link_a 20 family_a
            add_alternative link_b 10 family_a
            check_alternative link_a auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Families Priority Ascendant"
            add_alternative link_a 10 family_a
            add_alternative link_b 20 family_b
            check_alternative link_b auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Families Priority Descendant"
            add_alternative link_a 20 family_a
            add_alternative link_b 10 family_b
            check_alternative link_a auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Families Priority Ascendant Multiple"
            add_alternative link_a 10 family_a
            add_alternative link_b 20 family_a
            add_alternative link_c 30 family_b
            check_alternative link_c auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Families Remove Manually Set"
            add_alternative link_a 10 family_a
            set_alternative link_a
            add_alternative link_c 30 family_b
            remove_alternative link_a
            check_alternative link_c auto
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Families Remove Link After Manually Set Multiple"
            add_alternative link_a 10 family_a
            set_alternative link_a
            add_alternative link_b 20 family_a
            add_alternative link_c 30 family_b
            remove_alternative link_a
            check_alternative link_b manual link_c
            clean_dir
        rlPhaseEnd
            
        rlPhaseStart FAIL "Family After Remove Manually Set"
            add_alternative link_a 10 ""
            set_alternative link_a
            add_alternative link_b 20 ""
            add_alternative link_c 30 family_a
            remove_alternative link_a
            check_alternative link_c auto
            clean_dir
        rlPhaseEnd

            ##########

        rlPhaseStart FAIL "Dynamic Follower Add"
            add_alternative link_a 10 ""
            add_follower link_a follower_a
            check_alternative link_a auto follower_a
            clean_dir
        rlPhaseEnd

        rlPhaseStart FAIL "Dynamic Follower Add Auto"
            add_alternative link_a 10 ""
            add_alternative link_b 20 ""
            add_alternative link_c 5 ""
            add_follower link_a follower_a
            add_follower link_b follower_b
            check_alternative link_b auto follower_b
            clean_dir
        rlPhaseEnd

        rlPhaseStart FAIL "Dynamic Follower Add Manual"
            add_alternative link_a 10 ""
            add_alternative link_b 20 ""
            add_alternative link_c 5 ""
            add_follower link_a follower_a
            add_follower link_b follower_b
            set_alternative link_a
            check_alternative link_a manual link_b follower_a
            clean_dir
        rlPhaseEnd

        rlPhaseStart FAIL "Dynamic Follower Add And Remove Leader"
            add_alternative link_a 10 ""
            add_alternative link_b 20 ""
            add_alternative link_c 5 ""
            add_follower link_a follower_a
            add_follower link_b follower_b
            remove_alternative link_b
            check_alternative link_a auto follower_a
            clean_dir
        rlPhaseEnd

        rlPhaseStart FAIL "Dynamic Follower Remove"
            add_alternative link_a 10 ""
            add_follower link_a follower_a
            remove_follower link_a follower_a
            check_alternative link_a auto EMPTY
            clean_dir
        rlPhaseEnd

        rlPhase
    done

    # Cleanup phase: Remove test directory
    rlPhaseStartCleanup
        rlRun "rmdir $altdir $testdir $admindir"
    rlPhaseEnd
rlJournalEnd

# Print the test report
rlJournalPrintText



ntsysv.c
/* Copyright 1997-2007 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */
#include <alloca.h>
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include <libintl.h>
#include <locale.h>
#include <newt.h>
#include <popt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#define _(String) gettext((String))
#define UNIT_FILE_MAX 256

#include "leveldb.h"

/* return 1 on cancel, 2 on error, 0 on success */
static int servicesWindow(struct service *services, int numServices, int levels,
                          int backButton, char *root_path) {
    newtComponent label, subform, ok, cancel;
    newtComponent *checkboxes, form, curr, blank;
    newtComponent sb = NULL;
    newtGrid grid, subgrid, buttons;
    char *states;
    int i, done = 0, update = 0, j;
    int width, height;
    struct newtExitStruct e;
    int count = 0;
    int last = 0;

    newtPushHelpLine(_("Press <F1> for more information on a service."));

    newtGetScreenSize(&width, &height);
    width = width > 80 ? width - 60 : 20;
    height = height > 25 ? height - 17 : 8;

    sb = newtVerticalScrollbar(-1, -1, height, NEWT_COLORSET_CHECKBOX,
                               NEWT_COLORSET_ACTCHECKBOX);

    subform = newtForm(sb, NULL, 0);
    newtFormSetBackground(subform, NEWT_COLORSET_CHECKBOX);
    newtFormSetHeight(subform, height);

    checkboxes = alloca(sizeof(*checkboxes) * numServices);
    states = alloca(sizeof(*states) * numServices);

    for (i = 0; i < numServices; i++, count++) {
        if (last != services[i].type) {
            newtFormAddComponent(
                subform,
                newtCompactButton(-1, count,
                                  services[i].type == TYPE_INIT_D
                                      ? "SysV initscripts"
                                      : services[i].type == TYPE_XINETD
                                            ? "xinetd services"
                                            : services[i].type == TYPE_SYSTEMD
                                                  ? "systemd services"
                                                  : "Unknown"));
            count++;
            last = services[i].type;
        }
        if (services[i].type == TYPE_XINETD) {
            checkboxes[i] =
                newtCheckbox(-1, count, services[i].name,
                             services[i].levels ? '*' : ' ', NULL, states + i);
        } else if (services[i].type == TYPE_SYSTEMD) {
            checkboxes[i] =
                newtCheckbox(-1, count, services[i].name,
                             services[i].enabled ? '*' : ' ', NULL, states + i);
        } else {
            checkboxes[i] =
                newtCheckbox(-1, count, services[i].name,
                             (services[i].currentLevels & levels) ? '*' : ' ',
                             NULL, states + i);
        }
        newtFormAddComponent(subform, checkboxes[i]);
    }

    newtFormSetWidth(subform, width);

    buttons = newtButtonBar(_("Ok"), &ok, backButton ? _("Back") : _("Cancel"),
                            &cancel, NULL);

    blank = newtForm(NULL, NULL, 0);
    newtFormSetWidth(blank, 2);
    newtFormSetHeight(blank, height);
    newtFormSetBackground(blank, NEWT_COLORSET_CHECKBOX);

    subgrid =
        newtGridHCloseStacked(NEWT_GRID_COMPONENT, subform, NEWT_GRID_COMPONENT,
                              blank, NEWT_GRID_COMPONENT, sb, NULL);

    label = newtTextboxReflowed(-1, -1, _("What services should be "
                                          "automatically started?"),
                                30, 0, 20, 0);
    grid = newtGridBasicWindow(label, subgrid, buttons);

    form = newtForm(NULL, NULL, 0);
    newtGridAddComponentsToForm(grid, form, 1);
    newtGridWrappedWindow(grid, _("Services"));
    newtGridFree(grid, 1);

    newtFormAddHotKey(form, NEWT_KEY_F1);

    while (!done) {
        newtFormRun(form, &e);

        if (e.reason == NEWT_EXIT_HOTKEY) {
            if (e.u.key == NEWT_KEY_F12) {
                done = 1;
                update = 1;
            } else {
                /* must be F1 */
                curr = newtFormGetCurrent(subform);
                for (i = 0; i < numServices; i++)
                    if (curr == checkboxes[i])
                        break;

                if (i < numServices && services[i].desc)
                    newtWinMessage(services[i].name, _("Ok"), services[i].desc);
            }
        } else if (e.u.co == ok || e.u.co == cancel) {
            done = 1;
            update = (e.u.co == ok);
        }
    }

    newtPopWindow();
    newtFormDestroy(form);

    if (!update)
        return 1;

    for (i = 0; i < numServices; i++) {
        if (services[i].type == TYPE_XINETD) {
            if ((services[i].enabled && states[i] != '*') ||
                (!services[i].enabled && states[i] == '*'))
                setXinetdService(services[i], states[i] == '*', root_path);
        } else if (services[i].type == TYPE_SYSTEMD) {
            char *cmd = NULL;
            int en = 0;
            if (services[i].enabled && states[i] != '*')
                en = 0;
            else if (!services[i].enabled && states[i] == '*')
                en = 1;
            else
                continue;
            if(root_path) {
                asprintf(&cmd, "/usr/bin/systemctl --root=%s %s %s >/dev/null 2>&1",
                         root_path, en ? "enable" : "disable", services[i].name);
            } else {
                asprintf(&cmd, "/usr/bin/systemctl %s %s >/dev/null 2>&1",
                         en ? "enable" : "disable", services[i].name);
            }

            if (cmd == NULL)
                return 1;
            system(cmd);
            free(cmd);
        } else {
            for (j = 0; j < 7; j++) {
                if (levels & (1 << j))
                    doSetService(services[i], j, states[i] == '*', root_path);
            }
        }
    }

    return 0;
}

static int serviceNameCmp(const void *a, const void *b) {
    const struct service *first = a;
    const struct service *second = b;

    if (first->type != second->type) {
        return first->type - second->type;
    }

    return strcmp(first->name, second->name);
}

int getSystemdServices(struct service **servicesPtr, int *numServicesPtr, char *root_path) {
    FILE *sys = NULL;
    char service[UNIT_FILE_MAX + 1];
    int i;
    int r = 0;
    int c = ' ';
    char *cmd = NULL;
    struct service *services = NULL;
    struct service *p = NULL;
    int numServices = 0;
    int numServicesAlloced = 10;

    numServicesAlloced = 10;
    services = malloc(sizeof(*services) * numServicesAlloced);
    if (services == NULL)
        return -ENOMEM;

    if(root_path) {
        asprintf(&cmd, "systemctl --root=%s list-unit-files --no-legend --no-pager", root_path);
    } else {
        asprintf(&cmd, "systemctl list-unit-files --no-legend --no-pager");
    }
    sys = popen(cmd, "r");
    if (!sys) {
        r = -1;
        goto finish;
    }

    while (c != EOF) {
        int enabled;
        char *suffix;

        for (i = 0; (c = fgetc(sys)) != EOF && c != ' ' && i < UNIT_FILE_MAX;
             i++)
            service[i] = c;

        if (i == UNIT_FILE_MAX) {
            while ((c = fgetc(sys)) != '\n' && c != EOF)
                ;
            continue;
        }

        service[i] = '\0';
        if (c == EOF)
            break;

        while ((c = fgetc(sys)) == ' ')
            ;
        if (c == EOF)
            break;

        enabled = c == 'e' ? 1 : c == 'd' ? 0 : -1;
        while ((c = fgetc(sys)) != '\n' && c != EOF)
            ;

        if (enabled == -1)
            continue;

        if (strrchr(service, '@'))
            continue;

        suffix = strrchr(service, '.');
        if (suffix == NULL)
            continue;
        suffix++;
        if (strcmp(suffix, "service") && strcmp(suffix, "socket"))
            continue;

        if (numServices == numServicesAlloced) {
            numServicesAlloced += 10;
            p = realloc(services, numServicesAlloced * sizeof(*services));
            if (p == NULL) {
                r = -ENOMEM;
                goto finish;
            }
            services = p;
        }

        bzero(services + numServices, sizeof(struct service));

        services[numServices].name = strdup(service);
        if (services[numServices].name == NULL) {
            r = -ENOMEM;
            goto finish;
        }

        readSystemdUnitProperty(service, "Description",
                                &services[numServices].desc);

        services[numServices].enabled = enabled;
        services[numServices].type = TYPE_SYSTEMD;
        numServices++;
    }

    p = realloc(*servicesPtr,
                (*numServicesPtr + numServices) * sizeof(struct service));
    if (p == NULL) {
        r = -ENOMEM;
        goto finish;
    }
    *servicesPtr = p;
    memcpy(*servicesPtr + *numServicesPtr, services,
           numServices * sizeof(struct service));
    *numServicesPtr = *numServicesPtr + numServices;

finish:
    if (services)
        free(services);

    if (sys)
        pclose(sys);

    return r;
}

static int getServices(struct service **servicesPtr, int *numServicesPtr,
                       int backButton, int honorHide, char *root_path) {
    DIR *dir;
    struct dirent *ent;
    struct stat sb;
    char fn[1024];
    char *filename = NULL;
    struct service *services;
    int numServices = 0;
    int numServicesAlloced, rc;
    int err = 0;
    int systemd = systemdActive();

    numServicesAlloced = 10;
    services = malloc(sizeof(*services) * numServicesAlloced);

    if(root_path) {
        asprintf(&filename, "%s", root_path);
    } else {
        asprintf(&filename, RUNLEVELS "/init.d");
    }

    if (!(dir = opendir(filename))) {
        fprintf(stderr, "failed to open %s: %s\n", filename,
                strerror(errno));
        return 2;
    }

    while ((ent = readdir(dir))) {
        if (strchr(ent->d_name, '~') || strchr(ent->d_name, ',') ||
            (ent->d_name[0] == '.'))
            continue;

        if (systemd && isOverriddenBySystemd(ent->d_name))
            continue;

        if(root_path) {
            sprintf(fn, "%s/%s", root_path, ent->d_name);
        } else {
            sprintf(fn, RUNLEVELS "/init.d/%s", ent->d_name);
        }

        if (stat(fn, &sb)) {
            fprintf(stderr, _("error reading info for service %s: %s\n"),
                    ent->d_name, strerror(errno));
            continue;
        }
        if (!S_ISREG(sb.st_mode))
            continue;

        if (numServices == numServicesAlloced) {
            numServicesAlloced += 10;
            services =
                realloc(services, numServicesAlloced * sizeof(*services));
        }

        rc = readServiceInfo(ent->d_name, TYPE_INIT_D, services + numServices,
                             honorHide, root_path);

        if (!rc) {
            int i;

            rc = -2;
            for (i = 0; i < 7; i++) {
                if (isConfigured(ent->d_name, i, NULL, NULL)) {
                    rc = 0;
                    break;
                }
            }
        }

        if (rc == -1) {
            fprintf(stderr, _("error reading info for service %s: %s\n"),
                    ent->d_name, strerror(errno));
            continue;
        } else if (!rc)
            numServices++;
    }

    closedir(dir);

    if(root_path) {
        asprintf(&filename, "%s", root_path);
    } else {
        asprintf(&filename, "/usr/sbin/xinetd");
    }

    if (!stat(filename, &sb)) {
        if (!(dir = opendir(XINETDDIR))) {
            fprintf(stderr, "failed to open %s: %s\n",
                    filename, strerror(errno));
            return 2;
        }

        while ((ent = readdir(dir))) {
            if (strchr(ent->d_name, '~') || strchr(ent->d_name, ',') ||
                strchr(ent->d_name, '.'))
                continue;

            sprintf(fn, "%s/%s", (root_path) ? root_path : XINETDDIR, ent->d_name);
            if (stat(fn, &sb)) {
                err = errno;
                continue;
            }
            if (!S_ISREG(sb.st_mode))
                continue;

            if (numServices == numServicesAlloced) {
                numServicesAlloced += 10;
                services =
                    realloc(services, numServicesAlloced * sizeof(*services));
            }

            rc = readXinetdServiceInfo(ent->d_name, services + numServices, root_path);

            if (rc == -1) {
                fprintf(stderr, _("error reading info for service %s: %s\n"),
                        ent->d_name, strerror(errno));
                closedir(dir);
                return 2;
            } else if (!rc)
                numServices++;
        }

        if (err) {
            fprintf(stderr, _("error reading from directory %s: %s\n"),
                    (root_path) ? root_path : XINETDDIR, strerror(err));
            return 1;
        }
    }

    getSystemdServices(&services, &numServices, root_path);

    qsort(services, numServices, sizeof(*services), serviceNameCmp);

    *servicesPtr = services;
    *numServicesPtr = numServices;

    return 0;
}

int main(int argc, const char **argv) {
    struct service *services;
    int numServices;
    int levels = -1;
    char *levelsStr = NULL;
    char *root_path = NULL;
    char *progName;
    poptContext optCon;
    int rc, backButton = 0, hide = 0;
    struct poptOption optionsTable[] = {
        {"back", '\0', 0, &backButton, 0},
        {"level", '\0', POPT_ARG_STRING, &levelsStr, 0},
        {"root", '\0', POPT_ARG_STRING, &root_path, 0},
        {"hide", '\0', 0, &hide, 0},
        {0, 0, 0, 0, 0}};

    setlocale(LC_ALL, "");
    bindtextdomain("chkconfig", "/usr/share/locale");
    textdomain("chkconfig");

    progName = (char *)argv[0];

    if (getuid() != 0) {
        fprintf(stderr, _("You must be root to run %s.\n"), progName);
        exit(1);
    }

    optCon = poptGetContext(progName, argc, argv, optionsTable, 0);
    poptReadDefaultConfig(optCon, 1);

    if ((rc = poptGetNextOpt(optCon)) < -1) {
        fprintf(stderr, "%s: %s\n",
                poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
                poptStrerror(rc));
        exit(1);
    }

    if(root_path) {
        if (*root_path != '/') {
            fprintf(stderr, _("bad argument to --root\n"));
            exit(2);
        }
    }

    if (levelsStr) {
        levels = parseLevels(levelsStr, 0);
        if (levels == -1) {
            fprintf(stderr, _("bad argument to --levels\n"));
            exit(2);
        }
    }

    if (getServices(&services, &numServices, backButton, hide, root_path))
        return 1;
    if (!numServices) {
        fprintf(stderr, _("No services may be managed by ntsysv!\n"));
        return 2;
    }

    newtInit();
    newtCls();

    newtPushHelpLine(NULL);
    newtDrawRootText(0, 0, "ntsysv " VERSION " - (C) 2000-2001 Red Hat, Inc. ");

    if (levels == -1)
        levels = parseLevels("2345", 0);

    rc = servicesWindow(services, numServices, levels, backButton, root_path);

    newtFinished();

    return rc;
}
